### PR TITLE
update variable constraints dependency order

### PR DESCRIPTION
**Pull request by d-callan**
this should get us things like [yaxis], [xaxis], [overlay, facet] 
resolves https://github.com/VEuPathDB/EdaDataService/issues/190
depends on https://github.com/VEuPathDB/EdaCommon/pull/20

may also resolve https://github.com/VEuPathDB/EdaDataService/issues/213
and obviate the need for https://github.com/VEuPathDB/EdaDataService/issues/119

### CODE REVIEW

File: src/main/java/org/veupathdb/service/eda/ds/plugin/abundance/AbundanceBoxplotPlugin.java [hunk 1]
```diff
@@ -55,7 +55,7 @@ protected Class<AbundanceComputeConfig> getComputeSpecClass() {
   @Override
   public ConstraintSpec getConstraintSpec() {
     return new ConstraintSpec()
-      .dependencyOrder("overlayVariable", "facetVariable")
+      .dependencyOrder(new String[] {"overlayVariable", "facetVariable"})
```
Comment (ryanrdoherty) [inline src/main/java/org/veupathdb/service/eda/ds/plugin/abundance/AbundanceBoxplotPlugin.java:58]: @d-callan If you're willing to go back to all these, you can get rid of some boilerplate like this:
```
  String[][] blah = new String[][] {
    {"a","b"}, {"c","d"}
  };
```

**Review by ryanrdoherty — APPROVED** [review]

**Comment by d-callan** [comment]
@bobular maybe confirm the way ive set up the new dependency order makes sense to you as well? and works for front end dev? 

@ryanrdoherty how does merging this work given the edacommon dependency? i imagine i should include a version update for the dependency somewhere in this pr?

**Comment by ryanrdoherty** [comment]
@d-callan The steps would be to merge the EdaCommon PR to main and then release it, then update this PR with the new version.  But if you do that before the client is ready, we won't be able to add new stuff to main (e.g. the compute client stuff I'm working on).  Is the client stuff ready to go out too?

**Comment by d-callan** [comment]
@ryanrdoherty maybe I'm just being dense but why can't we add stuff to main in that case

**Comment by ryanrdoherty** [comment]
We can but it will break dev sites, right?

**Comment by bobular** [comment]
I'm just guessing after grepping some files, but would it be possible to add a temporary tag to the corresponding branch in https://github.com/VEuPathDB/EdaCommon/pull/20 - e.g. something like `v9.3.0pre`

Then update `build.gradle.kts` in this repo?
```
val edaCommon     = "9.3.0pre"         // EDA Common version
```

Then I might be able to get the `eda-data` service running locally via docker-compose magic?

**Comment by bobular** [comment]
Hi @d-callan 

I think there could be an issue with the array of arrays representation.

When we are talking about stratification variables, we want this:

`[["xAxisVariable"],["overlayVariable","facetVariable"]]`

to mean "overlay and facet variables are both constrained by xAxisVariable, but not by each other"

This is how I initially interpreted it.

But when I did the same for contingency tables, which have this dependency order:

`[["yAxisVariable","xAxisVariable"],["facetVariable"]]`

I constrained y and x variables only to be on the same branch as each other, but not depend on each other in any specific order.

So this would allow
```
y-axis: participant sex
x-axis: household water availability
```
which of course the current live site does using the old approach.

and also the other way around:
```
y-axis: household water availability
x-axis: participant sex
```
But the back end doesn't allow this. It complains about the variables being the wrong way round in the hierarchy.

Are we going to change the back end to allow both "directions" (x-y, y-x) for contingency tables?

If we don't, then I need a way to know that
``[["xAxisVariable"],["overlayVariable","facetVariable"]]`` doesn't mean `facetVariable` depends on `overlayVariable`
but in `[["yAxisVariable","xAxisVariable"],["facetVariable"]]` it DOES mean that `xAxisVariable` depends on `yAxisVariable`.

An alternative would be to use the old-style dependencies for cont tables, e.g.
`[["yAxisVariable"],["xAxisVariable"],["facetVariable"]]`

**Comment by d-callan** [comment]
I think the backed needs updating too.

**Comment by bobular** [comment]
OK, cool. I've just implemented it back to the my original interpretation and it seems to be working well for stratification. :smile:

**Review by bobular — APPROVED** [review]
All good. Don't forget to upgrade the "9.3.0pre" to, I guess, "9.3.0" when the other PR is merged.

**Pull request merged**